### ISSUE TITLE
FPKM transformation may be incorrect.

### ISSUE THREAD

**Issue by kfuku52** [kfuku52/amalgkit#32]
In [transcriptome_curation.r](https://github.com/kfuku52/amalgkit/blob/master/amalgkit/transcriptome_curation.r), I found that 1 is added to count values in FPKM calculation. This wouldn't be correct.
```
res = exp(log(counts + 1) - log(effective_lengths) - log(sum(counts)) + log(1e+09))
```

**Comment by Hego-CCTB** [comment]
This is a pseudocount to avoid undefined log values, in case of 0 counts.

**Comment by Hego-CCTB** [comment]
I probably could check any undefined values and set the counts to 0 and remove the +1 instead, if that is preferable.

**Comment by kfuku52** [comment]
If we intend this is FPKM, there shouldn't be a pseudocount. This is not a matter of preference, but of the definition. Also, please note that we don't have to worry about producing NAs if log/exp operations are removed. This is not a heavy calculation either.

**Comment by Hego-CCTB** [comment]
removed the +1, added a line to set inf values to 0

**Comment by kfuku52** [comment]
The pseudocount wasn't removed. Please double check.

**Comment by Hego-CCTB** [comment]
You are right. Must have slipped through during the merge. Should be fixed now!

**Comment by kfuku52** [comment]
Also, as I mentioned, log/exp should be removed. It only slows/complicates the program without any merits. We are dealing with the numbers that can be processed well with 64bit float, not something like 1e-300 FPKM. 

```
nrep = 200

counts = abs(rnorm(1e+5))
effective_lengths = as.integer(counts * 1000)

cat('FPKM with log/exp\n')
ptm = proc.time()
for (i in 1:nrep) {
    res = exp(log(counts) - log(effective_lengths) - log(sum(counts)) + log(1e+09))
    res[sapply(res, simplify = 'matrix', is.infinite)] <- 0
}
head(res)
proc.time() - ptm
cat('\n')

cat('FPKM without log/exp\n')
ptm = proc.time()
for (i in 1:nrep) {
    res = counts / effective_lengths / sum(counts) * 1e+09
}
head(res)
proc.time() - ptm
```

```
FPKM with log/exp
12.489479636025512.486604563248312.574810980266512.493770357616112.494973758743912.484778843257
   user  system elapsed 
 14.187   0.720  19.197 

FPKM without log/exp
12.489479636025512.486604563248312.574810980266512.493770357616112.494973758743912.484778843257
   user  system elapsed 
  0.396   0.061   0.522 
```

**Comment by Hego-CCTB** [comment]
yeah OK, calculating in log space is probably overkill. Should be fixed now.

I just noticed that I introduced a pseudocount to `tc` after fpkm/tpm calculation anyways.

     tc <- transform_raw_to_fpkm(tc, tc_eff_length)
     tc <- log(tc + 1)

shall I remove that as well?

**Comment by Hego-CCTB** [comment]
Did not adjust the formula for log space correclty. Fixed as well.

**Comment by kfuku52** [comment]
That is the log(N+1) transformation, so should be there. It's described in my paper.
https://www.nature.com/articles/s41467-020-18090-8

**Comment by kfuku52** [comment]
I am very concerned about the current situation where a lot of effort is required to fix a single line of a bug... Please check this commit https://github.com/kfuku52/amalgkit/commit/2d5acaf6c4239429e3d413dbeacfbc1d1eba1f3b.
